src/main/java/com/aws/greengrass/deployment/DeploymentConfigMerger.java

```
@@ -184,36 +181,61 @@ public static void waitForServicesToStart(Collection<GreengrassService> services
-                            DeploymentErrorCode.COMPONENT_BROKEN,
-                            DeploymentErrorCodeUtils.classifyComponentError(service, kernel));
-                }
-                if (!service.reachedDesiredState()) {
-                    allServicesRunning = false;
-                    continue;
-                }
-                if (State.RUNNING.equals(state) || State.FINISHED.equals(state) || !service.shouldAutoStart()
-                        && service.reachedDesiredState()) {
-                    continue;
-                }
+        while (!allServicesRunning(servicesToTrack, mergeTime, kernel)) {
+            Thread.sleep(WAIT_SVC_START_POLL_INTERVAL_MILLISEC); // hardcoded
+        }
+    }
+
+    /**
+     * Completes the provided future when all the listed services are running.
+     * Allows cancellation if the future is cancelled.
+     *
+     * @param servicesToTrack       services to track
+     * @param mergeTime             time the merge was started, used to check if a service is broken due to the merge
+     * @param kernel                kernel
+     * @param totallyCompleteFuture deployment result future, used to check if a deployment is cancelled
+     * @throws InterruptedException   if the thread is interrupted or deployment is cancelled while waiting here
+     * @throws ServiceUpdateException if a service could not be updated
+     */
+    public static void waitForServicesToStart(Collection<GreengrassService> servicesToTrack, long mergeTime,
+                                              Kernel kernel, CompletableFuture<DeploymentResult> totallyCompleteFuture)
+            throws InterruptedException, ServiceUpdateException {
+        while (!allServicesRunning(servicesToTrack, mergeTime, kernel)) {
+            if (totallyCompleteFuture.isCancelled()) {
+                throw new InterruptedException("Deployment is cancelled while waiting for services to start");
```

can we avoid using interrupts? Interrupt is low resolution. If we can throw an exception or return a value which clearly states that we're cancelled, then I'd really like that.